### ISSUE TITLE
v4.9.0 applies CVE-2021-45967 to hundreds of components

### ISSUE THREAD

**Issue by officerNordberg** [DependencyTrack/dependency-track#3146]
### Current Behavior

cpe matching got "fixed" and now CVEs with only vendors and no products erroneously match to everything ~~when fuzzy matching is enabled.  Fuzzy matching puts a wildcard in the~~ We receive sboms with * in vendor to deal with projects like apache http server and curl that have had CVEs filed under multiple vendors.    

### Steps to Reproduce

1. enable fuzzy matching
2. upload an sbom

### Expected Behavior

cpe's with an asterisk in the product position should not be fuzzy matched.

### Dependency-Track Frontend Version

4.8.x

### Browser

Google Chrome

### Operating System

Linux

### Checklist

- [X] I have read and understand the [contributing guidelines](https://github.com/DependencyTrack/dependency-track/blob/master/CONTRIBUTING.md#filing-issues)
- [X] I have checked the [existing issues](https://github.com/DependencyTrack/frontend/issues) for whether this defect was already reported

**Comment by nscuro** [comment]
You're saying we need an additional `&& !"*".equals(search.getProduct())` here?

https://github.com/DependencyTrack/dependency-track/blob/fe394cc254be02bfffc50f597e4f75bbe3611bcb/src/main/java/org/dependencytrack/search/FuzzyVulnerableSoftwareSearchManager.java#L128-L130

**Comment by officerNordberg** [comment]
So far, no, I don't think so.  I'm not even getting to the FuzzyMatcher based on this test.

```
    @Test
    public void testMyIssue() {
        qm.createConfigProperty(ConfigPropertyConstants.SCANNER_INTERNAL_FUZZY_ENABLED.getGroupName(), ConfigPropertyConstants.SCANNER_INTERNAL_FUZZY_ENABLED.getPropertyName(),"true", IConfigProperty.PropertyType.BOOLEAN, "");
        var project = new Project();
        project.setName("acme-app");
        project = qm.createProject(project, Collections.emptyList(), false);
        var component = new Component();
        component.setProject(project);
        component.setName("util-linux-setarch");
        component.setVersion("2.37.4");
        component.setCpe("cpe:2.3:a:*:util-linux-setarch:2.37.4:*:*:*:*:*:*:*");
        component = qm.createComponent(component, false);

        var vulnerableSoftware = new VulnerableSoftware();
        vulnerableSoftware.setCpe23("cpe:2.3:a:pascom_cloud_phone_system:*:*:*:*:*:*:*:*:*");
        vulnerableSoftware.setEdition("*");
        vulnerableSoftware.setPart("a");
        vulnerableSoftware.setProduct("*");
        vulnerableSoftware.setVersion("*");
        vulnerableSoftware.setVendor("pascom_cloud_phone_system");
        vulnerableSoftware.setVersionEndIncluding("7.19");
        vulnerableSoftware.setVulnerable(true);
        vulnerableSoftware = qm.persist(vulnerableSoftware);

        var vulnerability = new Vulnerability();
        vulnerability.setVulnId("CVE-2021-45967");
        vulnerability.setSource(Vulnerability.Source.NVD);
        vulnerability.setVulnerableSoftware(List.of(vulnerableSoftware));
        qm.createVulnerability(vulnerability, false);

        new InternalAnalysisTask().analyze(List.of(component));

        final PaginatedResult vulnerabilities = qm.getVulnerabilities(component);
        assertThat(vulnerabilities.getTotal()).isEqualTo(1);
        assertThat(vulnerabilities.getList(Vulnerability.class).get(0).getVulnId()).isEqualTo("CVE-2021-45967");
    }
```

I added some logging and it is getting the vsList filled here
https://github.com/DependencyTrack/dependency-track/blob/fe394cc254be02bfffc50f597e4f75bbe3611bcb/src/main/java/org/dependencytrack/tasks/scanners/InternalAnalysisTask.java#L142

I'm not sure what's going on in getAllVulnerableSoftware() but I see lots of changes since 4.8.2 https://github.com/DependencyTrack/dependency-track/compare/4.8.2...4.9.0

**Comment by officerNordberg** [comment]
Here's the issue. I don't know what this should be, but it certainly results in a weird "full outer join" of a match between ANY vendor and ANY product.  I can't think of any good solution to this other than a janky config or not adding * to the query as you move left to right in the parsedCpe if you've already encountered a '*'.  

` SELECT FROM org.dependencytrack.model.VulnerableSoftware WHERE ((part == '*' || part == :part) && vendor != null && (product == '*' || product == :product))`


https://github.com/DependencyTrack/dependency-track/blob/master/src/main/java/org/dependencytrack/persistence/VulnerableSoftwareQueryManager.java#L266

**Comment by Jasper-Ben** [comment]
:wave:

The current behavior "works as intended" (at least according to the CPE matching specification). When having a `*` in your components vendor CPE, everything else will match with the CVE CPE `cpe:2.3:a:pascom_cloud_phone_system:*:*:*:*:*:*:*:*:*` (which has `*` for basically every other attribute, so it matches with anything else your component CPE throws out there).

I am not sure how to fix this, without bringing back false negatives. Maybe an easy way of "clearing" a CVE from an entire project, not only on component basis would lessen the pain of dealing with this kind of situation?

**Comment by Jasper-Ben** [comment]
> We receive sboms with * in vendor to deal with projects like apache http server and curl that have had CVEs filed under multiple vendors.

The way CPE matching worked before v4.9.0, setting `*` as vendor would result in false negatives: https://github.com/DependencyTrack/dependency-track/issues/2580, unless the Fuzzy matching would by chance happen to "fuzz" all the correct CPEs, that is if fuzzy matching was enabled.

Basically, from a matching rule perspective this is a binary decision. We are stuck between hiding potential security risks by not treating `*` in CVE CPE as a match for a string `n` in component CPE (as was done < v.4.9.0) or having false positives when component CPE data is given that is incomplete. I for one prefer the latter and instead look for ways to better deal with the odd false positive.

> We receive sboms with * in vendor to deal with projects like apache http server and curl that have had CVEs filed under multiple vendors.

For this kind of use-case would it be useful to have CPE "aliases" fields in CycloneDX? So instead of throwing in a wildcard for covering the handful of vendors, you could list them explicitly? And then have all the aliases checked against the CVE CPE. @nscuro what do you think?

**Comment by nscuro** [comment]
> @Jasper-Ben For this kind of use-case would it be useful to have CPE "aliases" fields in CycloneDX? So instead of throwing in a wildcard for covering the handful of vendors, you could list them explicitly? And then have all the aliases checked against the CVE CPE.

FWIW CycloneDX intentionally only allows for a single CPE, as having more than one makes the component's identity ambiguous. I'd also think that, if you can't pinpoint the vendor, you'd likely need to use `ANY` anyway, as the NVD is notorious for inconsistent vendor naming. Choosing a hand full is a big gamble.

I think @officerNordberg has a point, and I'm wondering if there are legitimate cases where we'd want a match in these scenarios:

| Source Vendor | Source Product | Target Vendor | Target Product | Note |
| :------------------- | :-------------------- | :------------------ | :------------------ | :------ |
| ANY | ANY | ANY | ANY | Might as well match the entire NVD |
| ANY | foo | foo | ANY | Intention on Source side seems to be "unpredictably many vendors, just match all as long as product is the same", whereas Target side is "anything from vendor foo". The latter being a data quality problem in the NVD IMO. |
| foo | ANY | ANY | foo | Same as above, but attributes flipped |

The matching spec only considers one CPE attribute at a time, but doesn't cover cases where multiple attributes together don't really make sense. We could add some "common sense" filter (perhaps optional):

* If the target CPE has `vendor = *`, then don't match any source CPEs with `product = *`
* If the target CPE has `product = *`, then don't match any source CPEs with `vendor = *`
* If the target CPE has `vendor = *` and `product = *`, then ???

**Comment by Jasper-Ben** [comment]
> FWIW CycloneDX intentionally only allows for a single CPE, as having more than one makes the component's identity ambiguous.

Hm... I would argue that having a single CPE with an additional wildcard is more ambiguous:

```
{
  cpe: "cpe:2.3:a:*:curl:7.28.0:*:*:*:*:*:*:*"
}
```

than having multiple "explizit" CPEs that are "OR'd"

```
{
  "cpe": "cpe:2.3:a:curl_project:curl:7.28.0:*:*:*:*:*:*:*",
  "cpe_aliases": [
    "cpe:2.3:a:haxx:curl:7.28.0:*:*:*:*:*:*:*"
  ]
}
```

Personally, I think this would be the "cleaner" solution as it will still be closer to the NIST CPE matching specification "as it was intended".

> The matching spec only considers one CPE attribute at a time, but doesn't cover cases where multiple attributes together don't really make sense. We could add some "common sense" filter (perhaps optional):

    If the target CPE has vendor = *, then don't match any source CPEs with product = *
    If the target CPE has product = *, then don't match any source CPEs with vendor = *
    If the target CPE has vendor = * and product = *, then ???

That being said, I do not oppose this either. The NIST specification explicitly states that adding meaning to value combinations and therefore deciding what a "match" is, is a legitimate thing to do:

![image](https://github.com/DependencyTrack/dependency-track/assets/12223583/98924046-c5ab-4a64-b892-19a1633947f3)

I am only concerned that we might end up going down the wrong path again once we start "redefining" what a match is, overlooking legitimate positive matches in the process.

**Comment by nscuro** [comment]
> Personally, I think this would be the "cleaner" solution as it will still be closer to the NIST CPE matching specification "as it was intended".

I'd encourage you to bring up the possibility of CPE aliases for discussion in the CycloneDX community: https://cyclonedx.org/about/participate/#communications

What you can do today, is provide additional CPEs via `properties` of the component. Syft is doing that today, but Syft also puts lots of garbage CPEs in there that would cause false positives left and right. 

The `properties` approach however also assumes that both the SBOM generator and the SBOM ingestor understand the same property.

> I am only concerned that we might end up going down the wrong path again once we start "redefining" what a match is, overlooking legitimate positive matches in the process.

Hard to predict of course, but we will definitely not make any changes to the logic again that we haven't thoroughly tested and evaluated. If we end up customizing what a match is, the customized logic should most likely be behind a feature toggle. Definitely not willing to pay the price of false negatives again.

**Comment by Jasper-Ben** [comment]
> I'd encourage you to bring up the possibility of CPE aliases for discussion in the CycloneDX community: https://cyclonedx.org/about/participate/#communications

Will do :+1: 

A bit OT for this issue, but CPE aliases would also offer additional benefits.

For example, it allows for more traceability when vendor or product name changes, which is not a too uncommon thing in the open source world (e.g. by forking).

Lets take a look at terraform vs. opentofu as an example:

Terraform was forked at a specific version due to concerns around license changes. So if you decide to use the fork, you have an immediate change of vendor and product in your component CPE from `cpe:2.3:a:hashicorp:terraform:*:*:*:*:*:*:*:*` to something like `cpe:2.3:a:opentofu:opentofu:*:*:*:*:*:*:*:*` while most of your code base stays identical.

Now lets say a CVE is found in terraform that is reported by hashicorp as effecting everything up to `cpe:2.3:a:hashicorp:terraform:1.5.x:*:*:*:*:*:*:*`. Of course hashicorp will only report on their own product, not on any forks (why would they?). However, this particular CVE most likely will also affect opentofu (unless the offending code has been coincidentally reworked in the meantime). As a user of opentofu you now have to rely on someone actively monitoring and re-reporting on CVEs in the code shared with terraform (with a large project such as opentofu, this might not be an issue, but with smaller projects it surely would).

With cpe_aliases you could instead opt to add the latest version of the original product before the fork as a CPE alias:

```
{
  "cpe": "cpe:2.3:a:opentofu:opentofu:1.8.x:*:*:*:*:*:*:*",
  "cpe_aliases": [
    "cpe:2.3:a:hashicorp:terraform:1.5.x:*:*:*:*:*:*:*"
  ]
}
```

Of course, the rate of false positives would increase, the more code is changed since the fork, but it would prevent hiding any vulnerabilities reported to the original project that were introduced before the fork.

**Comment by officerNordberg** [comment]
> > FWIW CycloneDX intentionally only allows for a single CPE, as having more than one makes the component's identity ambiguous.
> 
> Hm... I would argue that having a single CPE with an additional wildcard is more ambiguous:
> 
> ```
> {
>   cpe: "cpe:2.3:a:*:curl:7.28.0:*:*:*:*:*:*:*"
> }
> ```
> 
> than having multiple "explizit" CPEs that are "OR'd"

That's impractical. Now I need to know all the aliases in order to match and magically add them somehow to my cyclonedx generation tool from self updating database.  We'd always be missing CVEs to start and literally have to be manually keeping up with CVEs ourselves in order to make sure DT matched them.

I had our DT all tweaked out prior to this update.  I feel like based on where this conversation is going I either need to fork off and die or come up with some new silly workaround to handle the case of when "random linux library/tool does not have a CVE yet to know exactly what vendor it's going to use, likely candidates being gnu, itself, itself_project from there it's a SWAG.  so in my cpe just make up a vendor, but definitely don't make it * but make fuzzy matching just try * after there's a miss on the original vendor."

**Comment by nscuro** [comment]
> I feel like based on where this conversation is going I either need to fork off and die or come up with some new silly workaround to handle the case [...]

I don't think this is a fair statement. We are more than happy to accommodate for your case, but we cannot simply return to the previous state because that simply implemented CPE matching wrong.

**Comment by Jasper-Ben** [comment]
> > than having multiple "explizit" CPEs that are "OR'd"
> 
> That's impractical. Now I need to know all the aliases in order to match and magically add them somehow to my cyclonedx generation tool from self updating database. We'd always be missing CVEs to start and literally have to be manually keeping up with CVEs ourselves in order to make sure DT matched them.
> 
> I had our DT all tweaked out prior to this update. I feel like based on where this conversation is going I either need to fork off and die or come up with some new silly workaround to handle the case of when "random linux library/tool does not have a CVE yet to know exactly what vendor it's going to use, likely candidates being gnu, itself, itself_project from there it's a SWAG. so in my cpe just make up a vendor, but definitely don't make it * but make fuzzy matching just try * after there's a miss on the original vendor."

I just want to emphasize again that the way matching was done pre-4.9.0 WILL cause false negatives in your use-case. I am pretty sure that if you would manually check said libraries that have a `*` as vendor, you would also find cases of false negatives, as DT would previously consider it a no-match if the component CPE attribute contains a `*` and the CVE CPE contains a value.

**Comment by Jasper-Ben** [comment]
Any case of "correct match" you would have had previously here, would be due to fuzzy matching, so based on "luck" and "magic" fuzzing algorithms without any consistency.

**Comment by officerNordberg** [comment]
@nscuro  I may have understated the willingness to improve the situation on your part, but it really feels like what was working for me and what @Jasper-Ben insists is right are at an impasse.  I don't see a workable solution. I used to think aliases was the way to go but as I pointed out, the unknowns that I'm trying to cover need to be explicitly known for that to work.  And yes, fuzzy matching is what I was relying on but now 4.9.0 supersedes the fuzzy matcher with it's own wildcard matching.

**Comment by Jasper-Ben** [comment]
> what @Jasper-Ben insists is right [...]

The part "I insists is right" is what is stated in the standard, which is what was fixed in v4.9.0. Sorry, but I won't let myself be "vilified" for pointing out that DT matching didn't conform to the standard. I am sorry, that this change made things worse for your specific case, I truly am. But standards exist for a reason. And in this case the non-alignment of DT caused valid vulnerabilities to be hidden from users, as reported in multiple issues by various people. And those were the ones "lucky" enough to notice it. The majority of users would have relied on DT to give them accurate results.

And as I stated before, I am also not strictly against @nscuro suggestion of adding custom definitions on what a match is (within the boundries of the standard), as long as all possible negative effects are ruled out.
In fact, we would also benefit from this, as we experience the same issue you describe. Currently we deal with this by applying VEX documents.

I just wanted to voice another possible option that might achieve the desired result without adding a custom definition of "match". If that is impracticle, sure let's look at other options.
